blog/2023-02-02-good-cognitive-friction.md

```diff
@@ -0,0 +1,251 @@
+> `MutArray<T>`) or snake (`array<T>`, `mut_array<T>`). Let us know what you think!
+
+Yes! We are going to make this *slightly* harder to define mutable collections. 
+
+> In the *future*, maybe we will introduce some syntactic sugar like:
+> 
+> ```js
+> let x = mut [1,2,3]; // <-- not a doctor
+> ```
+
+This design concept is what's called "good cognitive friction" (or "mechanical sympathy"). It is
+introduced intentionally in order to make sure the user understands the system better and encourage
+best practices.
+
+## Reassignability
+
+But immutability is not enough! Since we reference our array through `my_array`, the compiler also
+needs to guarantee that `my_array` will always point to the same object.
+
+Let's look at a hypothetic example:
+
+```js
+let i = 10;
+new cloud.Function(inflight () => { print(i); }) as "f1";
+i = 20;
+new cloud.Function(inflight () => { i = i + 9; }) as "f2";
+i = i - 90;
+```
+
+What value will the cloud function print? We can't tell because `i` is reassigned in multiple
+locations and there is absolutely no way to determine its value.
+
+This is where **reassignability** comes into play. In fact, in Wing, the above example would have
+failed compilation:
+
+```js
+   i = 20;
+// ^ variable i is not reassignable
+```
+
+OK, now we can relax. The Wing compiler tells us that `i` is not reassignable.
+
+Reassignability is a form of immutability (it is mutating the reference) and most modern programming
+languages are trying to encourage single assignment. `let` in Rust, `:=` in Go using `const`
+everywhere in JavaScript.
+
+So how do you make something reassignable? You can use `let var`:
+
+```js
+let var s = "hello";
+s = "world";
+```
+
+You can also use `var` in class/resource declarations:
+
+```js
+class Foo {
+  i: num;
+  var s: str;
+
+
+  init() {
+    // all non-optional fields must be assigned at construction (not implemented yet)
+    this.i = 10;
+    this.s = "world";
+  }
+
+  bar() {
+    // "var" fields can be reassigned at any time
+    this.s = "hello";
+
+    this.i = 20;
+    //   ^ i is not reassignable
+  }
+}
+```
+
+It can also be used in argument declarations:
+
+```js
+let handler = inflight (var x: str) => {
+  if x == "hello" {
+    x = "${x} world";
+  }
+};
+```
+
+## Why `let var`?
+
+We originally [considered](https://github.com/winglang/wing/pull/1180) using `var` instead of `let
+var`, but we realized this is making it too easy to folks to for the wrong thing. Entire code bases
+will be written with just `var` and mountains of linters will be written to protect you from
+shooting yourself in the foot.
+
+Going back to this concept of "good cognitive friction". If you need to type a few more characters in
+order to make a variable reassignable (`let var` versus `let`), you will likely just use `let` most
+of the time, and the world will be a better place with less bugs and happier developers.
+
+## The inflight Connection
```

Comment:
I would either capitalize "inflight" or put it inside backticks to emphasize it as a keyword

Reply:
Of course